daemon/oci_linux.go

```diff
@@ -179,7 +179,13 @@ func WithApparmor(c *container.Container) coci.SpecOpts {
 				s.Process = &specs.Process{}
 			}
 			s.Process.ApparmorProfile = appArmorProfile
+		} else {
+			// If AppArmor is not supported but a profile was specified, return an error
+			if c.AppArmorProfile != "" {
+				return errors.New("AppArmor is not supported on this host, but the profile '" + c.AppArmorProfile + "' was specified")
+			}
 		}
```

Comment:
I want to have a bit of a look if these are the most suitable places to put these and/or it we should validate this situation earlier on; at least I noticed that this code is used when _starting_ the container; it's used in `daemon.createspec` https://github.com/moby/moby/blob/32a29bfa95fecd02b4ba84a088167fb9cd7e6c1d/daemon/oci_linux.go#L1046
which is called when _starting the container; https://github.com/moby/moby/blob/32a29bfa95fecd02b4ba84a088167fb9cd7e6c1d/daemon/start.go#L143


I wonder if we should have a check for this when _creating_ the container; either in the router; https://github.com/moby/moby/blob/32a29bfa95fecd02b4ba84a088167fb9cd7e6c1d/api/server/router/container/container_routes.go#L466

Or as part of the daemon.CreateContainer, which already performs some validation steps;
https://github.com/moby/moby/blob/32a29bfa95fecd02b4ba84a088167fb9cd7e6c1d/daemon/create.go#L59-L79


I created a quick patch to always return an error, and while it is returned;

```bash
docker run -dit busybox
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
835f85a6d665: Download complete
Digest: sha256:c230832bd3b0be59a6c47ed64294f9ce71e91b327957920b6929a0caa8353140
Status: Downloaded newer image for busybox:latest
4ec5bfdd753a1be062cfed25679e80e338fb742b22c74b794c51420382592121
docker: Error response from daemon: FAILING HERE.
```

It's returned _after_ the container is already created, so even though the `run` failed, the non-functional container is still created, so now (as a user), I'd have to remove it;

```bash
docker ps -a
CONTAINER ID   IMAGE     COMMAND   CREATED          STATUS    PORTS     NAMES               SIZE
4ec5bfdd753a   busybox   "sh"      45 seconds ago   Created             heuristic_burnell   4.1kB (virtual 4.21MB)
```

Creating a container without starting won't show that something's wrong in the config until it's started;

```bash
docker create --name foo busybox
a650634608b961eab47e1e932f0106bcba503fd6f06543ce7598f93f7c5ffdf2

docker start foo
Error response from daemon: FAILING HERE
Error: failed to start containers: foo
```

Reply:
On a sidenote; I notice that BuildKit also can set an AppArmor profile, but it uses containerd's implementation for this, which at least doesn't have this check; https://github.com/moby/moby/blob/32a29bfa95fecd02b4ba84a088167fb9cd7e6c1d/vendor/github.com/moby/buildkit/executor/oci/spec_linux.go#L75

Reply:
Should I create a new PR on the moby/buildkit repo?